### ISSUE TITLE
Merge the platform and UI threads

### ISSUE THREAD

**Issue by knopp** [flutter/flutter#150525]
### Update 5/20/2025

As of Flutter 3.32 stable, threads are merged by default on iOS and Android with a flag available to opt-out. In Flutter 3.33 beta changes have landed so that threads will also be merged by default on Windows and macOS. Work to merge threads on Linux is WIP. We expect that the ability to opt-out of merged threads will be removed in a later release.

Known issues:

* [x] https://github.com/flutter/flutter/issues/163064
* [x] https://github.com/flutter/flutter/issues/175135
* [x] https://github.com/flutter/flutter/issues/173843

___

Currently Flutter has a separate UI thread (that runs dart code) and platform thread. I *think* the main purpose of the UI thread is  to prevent blocking the platform thread, but the actual benefit of this eludes me. In most applications the platform thread stays idle apart from tiny amount of work during event processing and compositing. The work is moved to UI thread, but if the work doesn't complete in time, it causes jank just like it would when blocking the platform thread. 

The drawbacks however seem pretty clear. The vast majority of platform APIs need to be accessed on platform thread. This means that instead of being able to call the API from Dart directly (i.e. through FFI), platform channels or something similar need to be used to marshal the call to platform thread. This adds some friction and asynchronous behavior to what could otherwise be simple synchronous call.

Situation gets even more complicated the other way around - when needed to call into Dart from platform thread. Platform callbacks are often synchronous (i.e. accessibility, event handling, IME, application delegate, and many more). Traditionally the approach here was to push enough state to platform thread to be able to provide synchronous response, but that's not always feasible or reliable enough. Eventual consistency fails in some cases, for example we're getting bitten by issue where IME on iOS generate fast sequence of events and then reads the text before the events have been processed by UI thread and sent back to platform thread.

On iOS for keyboard handling we need synchronous response, and to get it we keep pumping the CFRunLoop inside the event handler until we get it. This is a feasible approach for the problems above, but it's a tricky solution because of added reentrancy (the system code invoking the handler may not expect the runloop to be running while waiting for response) and edge cases (what if the isolate gets terminated while waiting for response?). With enough effort these are fixable (i.e. platform channels can have custom run loop mode (iOS, macOS) or custom execution so that it doesn't interfere with system run loop), but still, we're solving problem that probably shouldn't have existed in the first place.

### Non-exhaustive list of issues that would improve with merged UI and Platform thread
- Calling platform APIs directly and synchronously from Dart without needing platform channels or [platform isolate](https://github.com/flutter/flutter/issues/136314)
- Ability to get synchronous response from Dart code for platform callbacks, which would help with
    - accessibility 
    - event handling (keyboard events, drag & drop)
    - IME (cursor position query, no need to replicate composited text, could properly support custom bindings in `DefaultKeyBinding.dict`, ...)
    - Ability to synchronously respond to platform callbacks (i.e. window draggable area, `NSApplicationDelegate` methods)

### Possible drawbacks
- Non trivial change, would likely impact ecosystem (some plugins may need to get adapted, though most of plugins might just work if we keep platform channels working asynchronously)
- Possible performance regression with thread merging and shared raster/platform thread (linux). I think the proper solution here is to get rid of thread merging (adopt solution similar to what macOS is doing on all platforms) and implement separate raster thread in Linux embedder.

**Comment by jonahwilliams** [comment]
We'd almost certainly need a way to do this gradually and per platform. That would probably limit us from improving any APIs until after the transition had been completed.

The biggest question for me is what we can do about platform views. On iOS, I believe we can do some trickery to separate platform view updates from rasterization.

But Android is a bit of a mess. With TLHC we can manipulate platform views on a different thread, but with HC we have to manipulate them on the same thread. Because of the threading restrictions on older android devices + OpenGLES I dont know that we have a good thread merging solution that works on all android versions.

**Comment by stuartmorgan** [comment]
> * Non trivial change, would likely impact ecosystem (some plugins may need to get adapted, though most of plugins might just work if we keep platform channels working asynchronously)

I would expect essentially no impact on plugins using platform channels as long as we keep the default async, the real potential problem would be with packages using FFI where the threading of the code being called by FFI would be changed, without us being able to do anything about it centrally.

**Comment by mkustermann** [comment]
> This means that instead of being able to call the API from Dart directly (i.e. through FFI) ... need to be used to marshal the call to platform thread

For now we have the new [runOnPlatformThread](https://api.flutter.dev/flutter/dart-ui/runOnPlatformThread.html) API - which runs dart code in a stateful isolate on the platform thread. One can synchronously call into platform APIs and (AFAIK) platform APIs should be able to call synchronously back into Dart (via callbacks created using e.g. [NativeCallable.isolateLocal](https://api.dart.dev/stable/3.4.4/dart-ffi/NativeCallable/NativeCallable.isolateLocal.html)). 

That allows moving arbitrary dart logic to the platform thread. Though if the logic has to influence the UI then it has to interact asynchronously via the UI isolate - only merging the threads will solve this.

**Comment by johnmccutchan** [comment]
We plan on tackling this in the next year.

**Comment by knopp** [comment]
> For now we have the new [runOnPlatformThread](https://api.flutter.dev/flutter/dart-ui/runOnPlatformThread.html) API - which runs dart code in a stateful isolate on the platform thread. One can synchronously call into platform APIs and (AFAIK) platform APIs should be able to call synchronously back into Dart (via callbacks created using e.g. [NativeCallable.isolateLocal](https://api.dart.dev/stable/3.4.4/dart-ffi/NativeCallable/NativeCallable.isolateLocal.html)).

Indeed. I mentioned platform isolate in the original comment, but I do think that this is trying to solve a problem that shouldn't exist in the first place.

> But even if threads were merged, a call from the platform thread into Dart doesn't directly draw a new frame and therefore only affect UI asynchronously - as flutter engine determines when it's time to draw the next frame.

I don't think that's a problem (or even expected behavior). Not all callbacks from platform should trigger a frame (i.e. IME update, query draggable areas, etc).

**Comment by jonahwilliams** [comment]
Another example of a problematic API is semantics. The platform thread will query the flutter semantics tree at a particular rate, but since we don't know what that rate is we need to produce a full semantics tree on every frame.

**Comment by justinmc** [comment]
> (@jonahwilliams in https://github.com/flutter/flutter/issues/150460): FYI @knopp I'm trying to collect sets of issues that would be improved by merged platform/ui thread

Possibly the handling of large amounts of text in text fields. Currently the entire contents of the field get synced between the framework and engine on each keystroke (though this could possibly be mitigated using DeltaTextInputClient). https://github.com/flutter/flutter/issues/90063, https://github.com/flutter/flutter/issues/96173

**Comment by hellohuanlin** [comment]
To provide some context in iOS, both `UIKit` and `SwiftUI` runs on main thread. The only Apple framework I know that runs UI on a separate thread is `SceneKit` for 3D games. (SpriteKit for 2D games also runs UI on main thread).

**Comment by jonahwilliams** [comment]
@knopp I have a patch that adds a merged mode for iOS/android. If you were interested in looking at the desktop platforms, then some guidance on how window resizing interacts with threads would be appreciated.

**Comment by knopp** [comment]
On macOS raster thread is never merged but during resizing we block the platform thread until the frame is complete. Similar on windows and to a degree on Linux. We could pump the event loop instead of blocking, but that creates unexpected reentrancy in cocoa and can cause errors (i.e. events delivered while the resizing handler is active). This will definitely need some thought. In general the framework producing frame is completely synchronous process, so it should be possible to get this working without having to run the event loop.

**Comment by knopp** [comment]
I didn't notice your PR. I will play with it a little bit.

**Comment by delfme** [comment]
Can this help solve https://github.com/flutter/flutter/issues/116836
and https://github.com/flutter/flutter/issues/139230
?

**Comment by jonahwilliams** [comment]
This is done for iOS (Impeller) and Android. This quarter we'll investigate embedder changes necessary to bring this for desktop. Changes to iOS (Skia) are not planned as we intend to remove this backend sometime in the next few releases.

**Comment by knopp** [comment]
Desktop embedders are blocked on https://github.com/flutter/flutter/issues/152337. Once microtask queue pumping is resolved I can revive https://github.com/flutter/engine/pull/53747. Similar approach should work for Windows and Linux.

**Comment by jonahwilliams** [comment]
I do plan to look at that this quarter.

**Comment by bing127** [comment]
> I do plan to look at that this quarter.

您好，这个问题处理了么？何时处理呢

**Comment by eseidel** [comment]
Is there a design doc I can read to learn more modern context?  I didn't find one with https://github.com/flutter/flutter/issues?q=is%3Aopen+is%3Aissue+label%3A%22design+doc%22 but I'm also not sure if I'm looking the right way.  Thanks!

**Comment by knopp** [comment]
@eseidel, thanks for the feedback! I'll try to address your points:

iOS doesn't just have a separate compositing thread, it actually have a separate compositing process :) The big reason why iOS applications felt smooth compared to other platforms at the time was that all the animations were running in a completely separate high priority compositor process independent on the main application. The big downside for this is that the host application had very little control of these animations which didn't allow for any interactivity. 

Nowadays, most of the animations on iOS are driven from the client application main thread, as the devices became fast enough that it no longer matters and the ability to interact with animations outweighs this greatly. 

Flutter already has a separate raster thread, which is analogue to core animation compositor, and that is not going anywhere (in fact I have a PR that adds separate raster thread on linux).

You keep mentioning scrolling, would you mind elaborating on that? Scrolling used to be special long time ago where it was accelerated in hardware by framebuffer blits. I still have memories of first webpages with static background being unusable because static background prevented this acceleration from kicking in. But that's much different from what have these days :)

Scrolling on iOS is done in platform thread, where it changes offset of core animation layers so that the compositor can recompose them at new positions. There is no separate "scrolling thread". Likewise on Flutter, scrolling is part of layout, it happens on UI thread, I'm not quite sure what a separate scrolling thread would do?

If you can do the layout in time on UI thread, you will likely by able to do it on platform thread as well. There is very little work the platform thread does in a common Flutter application - there is some input event processing and some talking to compositor, but that's about it. Yes, a misbehaving plugin or a super heavy unoptimized platform views will interfere with this. But I think the solution for this is to fix the plugin rather then prevent the whole world from using native platform APIs in bidirectional synchronous way.

Regarding jank. If a plugin does too much work on platform thread, you will jank the application even now. There's no way around it. Pretty much on every platform, even with separate UI thread, there is part of frame processing where we need to talk to system compositor and that requires scheduling things on platform thread. If this is not possible because platform thread is blocked, you will get jank. The idea that separate UI thread somehow prevents this doesn't work in practice.

Regarding browsers, I think the situation is wildly different. There is very little the javacsript in browser can gain from running on platform thread, the isolation is much stronger and the number of concurrent tabs / javascript event loops much higher. It's not really apples to apples comparison.

**Comment by knopp** [comment]
> But those 3rd party plugins won't necessary notice this jank they're causing (especially if they're only wrapping iOS/Android libraries) since in normal iOS/Android apps it's OK to let the main thread go out to lunch for a few frames to do something (since scrolling/animations are happening on a separate thread anyway).

This is very wrong I'm afraid. The only thing you wouldn't notice platform thread "going out to lunch for a few frames" on iOS is for simple `CAAnimation`s such as rotating spinner. For anything more substantial, such as scrolling, this would **absolutely** cause a jank. The scrolling is performend on platform thread, it has always been, and if platform thread won't update core animation layer offsets in time for every frame, you will have jank. There's no way around this.

**Comment by bing127** [comment]
> [@eseidel](https://github.com/eseidel)，感谢您的反馈！我会尽力解决您的问题：
> 
> iOS 不仅有单独的合成线程，它实际上还有一个单独的合成过程 :) 当时 iOS 应用程序与其他平台相比感觉流畅的主要原因是所有动画都在完全独立的高优先级合成器进程中运行，独立于主应用程序。这样做的一大缺点是主机应用程序对这些动画的控制很少，因此无法实现太多交互。
> 
> 巨大的转变发生在 iOS 7 中。如今，iOS 上的大多数动画都是由客户端应用程序主线程驱动的，因为设备速度已经变得足够快，这不再是问题，而与动画交互的能力大大超过了这一点。
> 
> Flutter 已经有一个单独的光栅线程，它类似于核心动画合成器，并且不会去任何地方（事实上我有一个 PR 在 Linux 上添加了单独的光栅线程）。
> 
> 您一直在提到滚动，您介意详细说明一下吗？很久以前，滚动曾经是一种特殊功能，当时它通过帧缓冲区块传输在硬件中加速。我仍然记得第一个带有静态背景的网页无法使用，因为静态背景阻止了这种加速。但这与我们现在的情况大不相同 :)
> 
> iOS 上的滚动是在平台线程中完成的，它会改变核心动画层的偏移量，以便合成器可以在新位置重新合成它们。没有单独的“滚动线程”。同样，在 Flutter 上，滚动是布局的一部分，它发生在 UI 线程上，我不太确定单独的滚动线程会做什么？
> 
> 如果您可以在 UI 线程上及时完成布局，那么您很可能也可以在平台线程上完成布局。在常见的 Flutter 应用程序中，平台线程所做的工作很少 - 有一些输入事件处理和一些与合成器的对话，但仅此而已。是的，行为不当的插件或超重未优化的平台视图会干扰这一点。但我认为解决方案是修复插件，而不是阻止全世界以自然的双向同步方式使用本机平台 API。如上所述，无法做到这一点会产生一些重大影响。
> 
> 而影响是真实的。例如，文本输入是 Flutter 中被抱怨最多的部分之一，而平台渠道试图融入同步平台 API 的异步性质是其中的很大一部分。
> 
> 关于卡顿：如果插件在平台线程上执行了太多工作，那么现在应用程序也会卡顿。这是无法避免的。几乎在每个平台上，即使有单独的 UI 线程，在帧处理过程中，我们都需要与系统合成器或垂直同步等待器进行通信，而这通常需要在平台线程上安排任务。如果由于平台线程被阻塞而无法做到这一点，就会出现卡顿。单独的 UI 线程以某种方式防止这种情况的想法在实践中行不通。
> 
> 至于浏览器，我认为情况完全不同。浏览器中的 javacsript 几乎无法从平台线程上运行中获益，隔离性更强，并发选项卡/javascript 事件循环的数量更高。这实际上并不是同类比较。

您好，我看到您有处理 Ui 线程方面的工作，很感谢您的付出，我个人也一直在关注这块，从去年六月份到今天，为此我专门把这个 issues 添加到书签，基本每天都会看，我想问现在处理了页面底部有输入框，类似聊天页面，键盘弹出安卓会有明显的卡顿，这个问题处理了么？主要是我这边有太多的客户反馈这个问题了😮‍💨

**Comment by stuartmorgan** [comment]
> One advantage of the message channels is portability.

We've already moved pretty strongly away from message channels as a cross-platform API surface. We've fully abandoned it in all of the plugins we maintain (as it was a maintenance nightmare) in favor of the cross-platform API being defined in Dart, and for embeddings new APIs get formal C APIs in embedder.h instead of poorly spec'd, poorly documented, and harder-to-use de-facto message channel APIs.

**Comment by eseidel** [comment]
I appreciate the responses.  Busy day.  I intend to reply with a doc when I have time.

**Comment by zqxiaojin** [comment]
For hybrid  flutter applications, switching threads to the platform's main thread is a disaster. 
This is because hybrid apps have both Flutter pages and native pages.

Here are some examples:
Example 1: Preloading Flutter pages.
When a native page needs to open a Flutter page, we can load the Flutter logic in a background thread and perform loading in parallel, which does not affect the platform's main thread.

Example 2: Opening a native page from a Flutter page.
At this point, the Flutter page acts like a background thread and does not interfere with the current main thread.

Reducing the number of threads is beneficial, but mobile devices already have sufficient threads. I hope to preserve the existing thread design to lessen the burden on developers.

**Comment by knopp** [comment]
@zqxiaojin, is this a theoretical concern or do you have an application that has regressed? Do you have a repro I can look at and profile?

1. Preloading. If preloading does too much work on platform thread _before reaching dart entry point_ this is something we could possibly improve by doing some of the work on background thread. If preloading does too much work on platform thread _after the dart entry point_ (user dart code) then the proper solution would be to run the work on background isolate perhaps?

2. Opening native page. This would mean that either Flutter part or Native part are doing too much work on the main thread. Yes, the amount of work they can do is cumulatively is less. Still it would be beneficial to know what exactly is going on each part to see where the bottle neck is.

**Comment by zqxiaojin** [comment]
@knopp 

Under normal circumstances, a large amount of Dart logic running in a background thread should not cause any stuttering, as there are many async/await calls to distribute the time slices.

The stuttering situations I encountered only occurred after frequent calls to the method channel or when transmitting a large amount of data. For hybrid apps, the issues are quite clear and can be optimized relatively easily.

If the thread model is modified to use the platform's main thread, then many Dart codes that originally ran in the background thread will need to be reviewed, and what was previously smooth may become stuttery.

**Comment by knopp** [comment]
> Under normal circumstances, a large amount of Dart logic running in a background thread should not cause any stuttering, as there are many async/await calls to distribute the time slices.
> 
> The stuttering situations I encountered only occurred after frequent calls to the method channel or when transmitting a large amount of data. For hybrid apps, the issues are quite clear and can be optimized relatively easily.
> 
> If the thread model is modified to use the platform's main thread, then many Dart codes that originally ran in the background thread will need to be reviewed, and what was previously smooth may become stuttery.

Again, the devil is in the details. What calls exactly? What data? One possible benefit of merging threads is is that you don't need to use platform channels anymore, you can make FFI calls instead. You might not need to transfer the data, you may be able to operate on the data directly. It really depends on specific situation, that's why I'd appreciate a concrete example.

**Comment by mit-mit** [comment]
As of Flutter 3.32 stable, threads are merged by default on iOS and Android with a flag available to opt-out. If anyone sees issues with this, please report them promptly (including details and repro steps).

**Comment by ManifoldWanderer** [comment]
I don't know if this is the right place to mention this, but in the example from the [What’s new in Flutter 3.32](https://medium.com/flutter/whats-new-in-flutter-3-32-40c1086bab6e) Medium post, to enable merged threads on Windows, you should use

`project.set_ui_thread_policy(flutter::UIThreadPolicy::RunOnPlatformThread);`

instead of the line without the `flutter::` namespace prefix, as it is now in the article, which will give you a compilation error like this: _error C2653: 'UIThreadPolicy': is not a class or namespace name_.

It might be worth updating the article with the corrected code to help others avoid this issue.

With this correction, it works great! For example, I can now resize the app window directly in Flutter code using the `win32` package, without diving into C++, which I think is a nice step ahead for Flutter Desktop:

```dart
void main() {
  WidgetsFlutterBinding.ensureInitialized();
  final flutterWindowHandle = FindWindow('FLUTTER_RUNNER_WIN32_WINDOW'.toNativeUtf16(), nullptr);
  MoveWindow(flutterWindowHandle, 0, 0, 1000, 500, TRUE);
  runApp(const MyApp());
}
```

**Comment by dickermoshe** [comment]
An unintended consequence of this PR is that current Flutter Plugins were not developed with the assumption that they would be on the main thread. Therefore, they perform blocking work on that thread. Usually, such work would be offloaded to another isolate, but at the time, developers didn't feel that was necessary because the plugins were not on the UI thread.

The release of 3.22 now causes many of these plugins to introduce jank on the main UI thread

**Comment by stuartmorgan-g** [comment]
> An unintended consequence of this PR is that current Flutter Plugins were not developed with the assumption that they would be on the main thread.

Plugins have always been created on the main thread, and message channel messages were always delivered on the main thread. If plugin authors were assuming otherwise, that has always been incorrect.

FFI should be the only case that changes, [as noted above](https://github.com/flutter/flutter/issues/150525#issuecomment-2179609763).

> The release of 3.22 now causes many of these plugins to introduce jank on the main UI thread

Could you provide specific examples? The only case I'm aware of where this change would cause existing plugins to jank would be:
- they use FFI rather than platform channels, and
- they were doing expensive work on the main Dart isolate, which would also always have been an bad idea, and something that would likely cause jank.

**Comment by bizz84** [comment]
@mit-mit 

> As of Flutter 3.32 stable, threads are merged by default on iOS and Android with a flag available to opt-out. If anyone sees issues with this, please report them promptly (including details and repro steps).

What flag can we use to opt-out? I can't find documentation about this anywhere.

I'll open a separate issue for some related performance problems I'm experiencing with the [`speech_to_text`](https://github.com/csdcorp/speech_to_text) package.

**Comment by CarGuo** [comment]
@bizz84  


iOS : FLTEnableMergedPlatformUIThread https://github.com/flutter/flutter/pull/162883 

Andorid:

```xml
        <meta-data
            android:name="io.flutter.embedding.android.DisableMergedPlatformUIThread"
            android:value="true" />
```

**Comment by mboetger** [comment]
Sorry, was in the wrong tab :( didn't mean to close this issue

**Comment by loic-sharma** [comment]
> It might be worth updating the article with the corrected code to help others avoid this issue.

Fixed! Thanks for reporting this :)

**Comment by orestesgaolin** [comment]
Can you clarify whether the thread merging was enabled for iOS/Android on previous versions of Flutter (e.g. 3.29?) or is it only enabled starting from 3.32?

**Comment by delfme** [comment]
Thread merging is not enabled in flutter v3.24.5

**Comment by SrAdam** [comment]
I've spoken with many developers in real life who rely on the flags, but I notice they often don't engage with the relevant GitHub issues.

I'm here to show my support and kindly request that the flag for maintaining the separate UI thread remain available at least through the end of 2025.

Many production apps still depend on this threading model for various reasons (heavy computations, architectural constraints, etc.), and removing it prematurely would force rushed refactoring that could introduce bugs.

Thank you for considering the needs of apps that require more time to adapt to the merged threading model.

**Comment by bizz84** [comment]
@SrAdam 

> I'm here to show my support and kindly request that the flag for maintaining the separate UI thread remain available at least through the end of 2025.

100% agree. We all know that in the open-source space, changes can take a long time. It would be unreasonable to force everyone to the new merged thread approach, without at least verifying that most popular plugins have migrated.

As I was mentioning in https://github.com/flutter/flutter/issues/170242, I still experience jank even with `FLTEnableMergedPlatformUIThread=false`, so I'm not sure the flag is even working properly as intended. I hope the team can look into this.

**Comment by stuartmorgan-g** [comment]
> It would be unreasonable to force everyone to the new merged thread approach, without at least verifying that most popular plugins have migrated.

Please re-read https://github.com/flutter/flutter/issues/170242#issuecomment-2960225889 and https://github.com/flutter/flutter/issues/170242#issuecomment-2966401838.

If you have any evidence that "most popular plugins" need to make any changes at all, please provide a detailed list.

**Comment by srmncnk** [comment]
Showing my support here as well. I'll link an issue I had posted a few weeks ago which describes exactly how this breaking change completely breaks our app without any chance of workarounds: https://github.com/flutter/flutter/issues/169339#issuecomment-2940316941

As for "most popular plugins", we can start with plugins that deal with foreground services and additional threading which support options that flutter does not support out of the box: [flutter_isolate](https://pub.dev/packages/flutter_isolate), [flutter_foreground_task](https://pub.dev/packages/flutter_foreground_task), [background_locator](https://pub.dev/packages/background_locator_2), [location](https://pub.dev/packages/location), [flutter_background](https://pub.dev/packages/flutter_background), [flutter_foreground_service](https://pub.dev/packages/flutter_foreground_service), [flutter_foreground_plugin](https://pub.dev/packages/flutter_foreground_plugin)

**Comment by knopp** [comment]
Spawning another engine because you actually need another engine (i.e. because the engine is unfortunately tied to an activity on android) seems like a valid use case, though I'd much rather see this resolved by shared flutter engine decoupled form the activity and a having a `FlutterView` per activity. But in any case, this part is completely orthogonal to threading.

Spawning another engine because you want to run things in background thread seems like a something that should not be necessary. One of the reason I kept seeing looking at the plugins is that background isolates can't interact with plugin API, but that seems bit strange because a) going through plugin API rethreads the call to main thread anyway and b) you can just use FFI/JNI for bidirectional communication while doing everything on the background thread.

**Comment by delfme** [comment]
@srmncnk please correct me if I’m misunderstanding. You’re describing a continously running isolate that does things in background; that isn’t very “mobile-friendly,” and iOS would likely kill it. Could you clarify why the app needs this approach?

Typically, mobile apps connect to a server while in the foreground and rely on periodic background tasks to either perform lightweight actions (e.g. fetching data from an HTTP endpoint) or handle longer operations like cloud backups. Off the top of my head, I’m not sure why your use case requires something different.

**Comment by orestesgaolin** [comment]
> Typically, mobile apps connect to a server while in the foreground and rely on periodic background tasks to either perform lightweight actions (e.g. fetching data from an HTTP endpoint) or handle longer operations like cloud backups.

On Android you can have foreground service that runs continuously e.g. when connected to a BLE accessory or when navigating with GPS. OS can still limit it a bit but in general it's long running background work, in our case even up to 5-7 days long. We considered spawning flutter engine for that but in the end decided to just write it natively because of the flutter overhead. I suppose @srmncnk decided to pursue this approach. 

On iOS it's also possible to have long running headless process with Bluetooth background state restoration which keeps the app running (without active ViewController) as long as the Bluetooth device emits notifications of given characteristics.

**Comment by delfme** [comment]
@orestesgaolin thanks, I'd go native for BLE as well.

**Comment by srmncnk** [comment]
@delfme @knopp @orestesgaolin Thank you for your replies. Let me address the most important points to us:

> Spawning another engine because you want to run things in background thread seems like a something that should not be necessary. One of the reason I kept seeing looking at the plugins is that background isolates can't interact with plugin API, but that seems bit strange because a) going through plugin API rethreads the call to main thread anyway and b) you can just use FFI/JNI for bidirectional communication while doing everything on the background thread.

@knopp Agreed, however as noted elsewhere, traditional use of plugins 'as-is' is not possible if plugins have callbacks from method channel (not FFI). You can call into method channel from a pure Dart isolate if you register the root token via BackgroundIsolateBinaryMessenger, you will however not be able to receive callbacks from method channel. Agreed, you would probably be able to do it via FFI and NativeCallable.isolateLocal (haven't tried yet), but for that one would need to fork and correct all the plugins that have native callbacks.

That's only one reason why one would want this (callbacks). The other reasons being: having a long living foreground service, surviving app kill and having a guarantee of running in a single(!) separate thread.

> @srmncnk please correct me if I’m misunderstanding. You’re describing a continously running isolate that does things in background; that isn’t very “mobile-friendly,” and iOS would likely kill it. Could you clarify why the app needs this approach?

@delfme Our app is a navigational logger app. We receive & log GPS locations in the background while navigating the user while in foreground. We don't mind ui being killed by the OS as long as the GPS locations are stored in background (long living foreground service). We also read a lot of sensors in the background, that's done very low level via FFI, for that having a separate thread is an absolute must, otherwise sensor data flow is compromised. We run crossplatform on Androd, iOS and our dedicated devices. iOS does not kill the isolate, app has been in the stores for 6 years now.

With this you can see that a lot of codebase is used in the long living isolate. With that in mind and with the plugins available at that time (even 6 years ago!) a prospect of having the same debuggable codebase on different platforms for a crucial part of the app outweights the overhead of having another few FlutterEngines running.

@orestesgaolin You're absolutely correct. At the design stage we ran a few battery consumption tests and decided that the overhead is well worth it.

**Comment by knopp** [comment]
> The other reasons being: having a long living foreground service, surviving app kill and having a guarantee of running in a single(!) separate thread.

If the long living foreground service is running in separate process (so that it can survive application crash), that means it has a dedicated main thread/looper which should be very much idle, so I don't quite see the problem with flutter engine running dart code on this thread?

**Comment by srmncnk** [comment]
> > The other reasons being: having a long living foreground service, surviving app kill and having a guarantee of running in a single(!) separate thread.
> 
> If the long living foreground service is running in separate process (so that it can survive application crash), that means it has a dedicated main thread/looper which should be very much idle, so I don't quite see the problem with flutter engine running dart code on this thread?

@knopp That's not entirely correct. The foreground service does not need to be running in a separate process in order to survive an application crash. What crashes (or is being killed) is the Activity to which the first, original FlutterEngine, is bound to. When you create another FlutterEngine within a foreground service and bind it to its context, it will survive. It will survive even if you kill the app while swiping it in the task manager on Android. It will even survive with merged platform & ui threads - so even if Dart code is running in same thread but different isolates, the second one - bound to foreground service context - will keep executing. Just tested that.

**Comment by rubenferreira97** [comment]
> Plugins have always been created on the main thread, and message channel messages were always delivered on the main thread. If plugin authors were assuming otherwise, that has always been incorrect.

It's clear that many plugin developers (myself included) have incorrect assumptions about threading architecture in Flutter. If this misunderstanding is widespread (as proven by many plugin implementations), it may indicate a gap in the official documentation or developer guidance. Could we improve the documentation or provide clearer best practices on threading and isolate behavior, especially in topics like `plugins`, `platform channels` and `ffi`? This could help avoid future issues and strengthen the overall ecosystem.

Currently, there are several native interop options that I am aware of: `platform channels`, `ffi`, `ffigen`, `jni`, `jnigen`, and `swiftgen` (though not yet published). To those less familiar, it's unclear which to use when and why, especially as many are still in beta or not fully production-ready.

Official documentation on use cases would be very helpful. One practical example that would benefit from clearer guidance is: how should a developer implement a foreground/background service that performs heavy computation via FFI without introducing jank? I’ve seen some example projects, but they are not very discoverable and often don’t explain in detail what, to me, are complex topics.

**Comment by knopp** [comment]
> @knopp That's not entirely correct. The foreground service does not need to be running in a separate process in order to survive an application crash. What crashes (or is being killed) is the Activity to which the first, original FlutterEngine, is bound to. When you create another FlutterEngine within a foreground service and bind it to its context, it will survive. It will survive even if you kill the app while swiping it in the task manager on Android. It will even survive with merged platform & ui threads - so even if Dart code is running in same thread but different isolates, the second one - bound to foreground service context - will keep executing. Just tested that.

There is no mechanism in posix I know of that allows "part of the process" to survive getting killed. The service manifest in android allows to run the service in separate process, which is how I assume it would survive killing the main process.

**Comment by srmncnk** [comment]
> > [@knopp](https://github.com/knopp) That's not entirely correct. The foreground service does not need to be running in a separate process in order to survive an application crash. What crashes (or is being killed) is the Activity to which the first, original FlutterEngine, is bound to. When you create another FlutterEngine within a foreground service and bind it to its context, it will survive. It will survive even if you kill the app while swiping it in the task manager on Android. It will even survive with merged platform & ui threads - so even if Dart code is running in same thread but different isolates, the second one - bound to foreground service context - will keep executing. Just tested that.
> 
> There is no mechanism in posix I know of that allows "part of the process" to survive getting killed. The service manifest in android allows to run the service in separate process, which is how I assume it would survive killing the main process.

Killing the process is the very last step in the lifecycle on Android - with that also foreground services (that do not run in a separate process) are killed. Firsty (and mostly) activities are killed. And since FlutterEngine is tied to either an Activity or a foreground service context, the latter will survive.

**Comment by knopp** [comment]
> > > [@knopp](https://github.com/knopp) That's not entirely correct. The foreground service does not need to be running in a separate process in order to survive an application crash. What crashes (or is being killed) is the Activity to which the first, original FlutterEngine, is bound to. When you create another FlutterEngine within a foreground service and bind it to its context, it will survive. It will survive even if you kill the app while swiping it in the task manager on Android. It will even survive with merged platform & ui threads - so even if Dart code is running in same thread but different isolates, the second one - bound to foreground service context - will keep executing. Just tested that.
> > 
> > 
> > There is no mechanism in posix I know of that allows "part of the process" to survive getting killed. The service manifest in android allows to run the service in separate process, which is how I assume it would survive killing the main process.
> 
> Killing the process is the very last step in the lifecycle on Android - with that also foreground services (that do not run in a separate process) are killed. Firsty (and mostly) activities are killed. And since FlutterEngine is tied to either an Activity or a foreground service context, the latter will survive.

(I edited the original reply but I'll put it here instead)

If only the activity is getting killed that is very different than "application" getting killed. It would bring down the entire flutter engine, but that's because of unfortunate decision to couple flutter engine with an activity. As I mentioned above, I think the ideal situation would be single flutter engine shared by all activities with a FlutterView per activity.

**Comment by srmncnk** [comment]
Flutter has one single Android activity, correct. But again, "application" getting killed on Android in most cases simply means that main activity is schedulled to die. Foreground services survive, they (and the whole process) get killed only when OS is so much without resources that it starts sending SIGKILLs.

There's a reason foreground services exist. And there's a reason foreground service plugins exist. We've been using this concept in our app now for the past 6 years and this, I can't put it otherwise, simply works.

**Comment by littleGnAl** [comment]
The Flutter engine performs GC when the view is destroyed,  in add-to-app scenario, which can easily cause latency. I don’t know why we need to perform GC at view destroyed — maybe it would be more reasonable to do it when the engine is destroyed, or add a flag to defer the cleanup until then.

<img width="1406" height="302" alt="Image" src="https://github.com/user-attachments/assets/440eb16a-bceb-4b0d-9945-86f1b3b7b9c4" />

**Comment by mraleph** [comment]
Yeah, I remember discussing this somewhere with Flutter engine team. This was added long time ago to keep memory under control in add2app scenarios where many Flutter views are created and destroyed in rapid succession. I think it does not scale to do stop-the-world GC in these situations especially now that threads are merged. (cc @chinmaygarde)

**Comment by chinmaygarde** [comment]
As of https://github.com/flutter/flutter/pull/170309, the engine no longer calls Dart_NotifyDestroyed.

**Comment by littleGnAl** [comment]
Another issue affecting developer experience, during a hot restart (likely related to JIT compilation), the app displays an ANR dialog, especially on the Android emulator.

**Comment by loic-sharma** [comment]
@littleGnAl could you file a new issue with reproduction steps?

**Comment by loic-sharma** [comment]
With merged threads, it’s important that plugins to do expensive work off the platform thread to avoid UI jank.

Currently, Flutter Windows plugins do not have an easy way to switch back to the platform thread. This makes it hard for Flutter Windows plugins to do work in a background thread. We should consider adding this API before we remove Flutter Windows’s merged threads opt-out. Tracked by: https://github.com/flutter/flutter/issues/79213

cc @knopp @mattkae

**Comment by littleGnAl** [comment]
I have another use case for a large Flutter project.
https://github.com/flutter/flutter/issues/79327#issuecomment-3223052705

**Comment by srmncnk** [comment]
Since Fluttercon @ Berlin is happening soon, is anybody going to be there, maybe we can have a chat about it?

**Comment by knopp** [comment]
> Since Fluttercon @ Berlin is happening soon, is anybody going to be there, maybe we can have a chat about it?

I will be there, happy to chat about this.

**Comment by mraleph** [comment]
> Since Fluttercon @ Berlin is happening soon, is anybody going to be there, maybe we can have a chat about it?

A bunch of Dart native interop people (@goderbauer @dcharkes @HosseinYousefi) and @mit-mit are attending - and I would expect that you should be able to meet them in the Flutter booth intermittently.

**Comment by srmncnk** [comment]
Thanks! I connected via LinkedIn, let's talk there. @knopp I'll be stopping by the flutter team for sure, looking forward to it.

**Comment by orestesgaolin** [comment]
Just to indicate how this change looks like between 3.19 and 3.35 - the same code running

```swift
import Flutter
import UIKit

@main
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    GeneratedPluginRegistrant.register(with: self)

    let controller : FlutterViewController = window?.rootViewController as! FlutterViewController

    let benchmarkChannel = FlutterMethodChannel(name: "com.example/benchmark", binaryMessenger: controller.engine!.binaryMessenger)
    benchmarkChannel.setMethodCallHandler { (call: FlutterMethodCall, result: @escaping FlutterResult) in
        if call.method == "blockMainThread" {
            DispatchQueue.main.async {
                let startTime = CFAbsoluteTimeGetCurrent()
                while CFAbsoluteTimeGetCurrent() - startTime < 10.0 {
                    // Busy wait - consumes CPU
                }
                result("Main thread blocked for 10 seconds")
            }
        } else if call.method == "blockPlatformThread" {
            // Execute on a background thread instead of the main thread
            DispatchQueue.global(qos: .background).async {
                let startTime = CFAbsoluteTimeGetCurrent()
                while CFAbsoluteTimeGetCurrent() - startTime < 10.0 {
                    // Busy wait - consumes CPU
                }
                // Return result on main thread to safely communicate back to Flutter
                DispatchQueue.main.async {
                    result("Platform thread blocked for 10 seconds")
                }
            }
        } else {
            result(FlutterMethodNotImplemented)
        }
    }
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
}
```

# 3.19.6

https://github.com/user-attachments/assets/5444859e-0db0-4171-b745-74d99f471388

# 3.35.4

https://github.com/user-attachments/assets/e0309e79-9d99-43ca-9665-b20ea75184b3



edit: "platform thread" should obviously be called "background thread"

**Comment by stuartmorgan-g** [comment]
Sure, but now add a third button at the bottom that has some visual effect (or even that does nothing, but has the ink tap effect). Is the experience of pressing that button right after pressing "Block Main Thread" good in 3.19? Or put these buttons in a scrollable context, and try to scroll in 3.19 after pressing "Block Main Thread".

The argument isn't that it's impossible to construct cases where specific things are worse now when plugins block the main thread for a long time, the argument is that it has *always been very bad* for plugins to block the main thread for a long time, and that the benefits of thread merging outweigh the downsides of making an anti-pattern slightly worse.

**Comment by orestesgaolin** [comment]
@stuartmorgan-g I hope my message above did not come across as an argument in favor of reverting this change or extending the opt-out period. I was experimenting with various Flutter versions and the opt-out flag to observe how things evolve over time. Video can say more than 100 messages in a thread ;)
Personally, I believe it would be better for the framework to merge the platform and UI threads, unlocking future possibilities. One argument I could make is that we should communicate more clearly with plugin authors, who may not be familiar with the threading limitations and realities. The change in the threading model requires them to reassess their code for cases like the one above. While some issues may be trivial to fix, others may not, and many authors may only realize this after upgrading the Flutter SDK.

**Comment by stuartmorgan-g** [comment]
> Video can say more than 100 messages in a thread ;)

And I'm pointing out that what those particular videos appear to say—because they don't show any attempt to interact with the app in any way after pressing the "Block Main Thread" button—is that a plugin blocking the main thread wasn't a problem in 3.19, which isn't actually true.

**Comment by orestesgaolin** [comment]
That's a good point! So to clarify, even in 3.19 blocking the main thread would lead to the native pointer events not getting delivered to the Flutter. Trying to scroll would be prevented, but taps would be queued and delivered with delay.

Nonetheless, the end-user experience would still be acceptable in most cases. In the problems we encountered, the `Thread.sleep` invocations used by one of our vendors were in the range of 100-200 ms. This delay could be noticeable during animations and navigation transitions, but it did not significantly affect the app's interactivity. (It's fixed now, though.)

https://github.com/user-attachments/assets/697b6280-6374-4d64-930a-1ffc71c9a1bb

**Comment by srmncnk** [comment]
> [@stuartmorgan-g](https://github.com/stuartmorgan-g) I hope my message above did not come across as an argument in favor of reverting this change or extending the opt-out period. I was experimenting with various Flutter versions and the opt-out flag to observe how things evolve over time. Video can say more than 100 messages in a thread ;)

Both points are valid and I think we can all agree that, moving forward, merged threads are the way to go to ensure interoperability and flexibility.

What I want to comment on, is the opt-out period. :) There are issues, where this became a major problem (https://github.com/flutter/flutter/issues/169339 in particular), but they relate to a headless flutter engine for background threading purposes. We've already had good talks about this yesterday and will continue to do so. Opt-out being possible for secondary non-main flutter engines would solve everything. For main flutter engine, there should be no doubt, that merged ui and platorm threads are the way to go. @knopp @dcharkes @orestesgaolin

**Comment by robert-ancell** [comment]
Proposing to change Linux default in https://github.com/flutter/flutter/pull/176759

**Comment by robert-at-pieces** [comment]
Since our upgrade to flutter 3.35.6 (from 3.32.6) our application is encountering many errors such as these on Windows:

```
[ERROR:flutter/shell/platform/windows/task_runner_window.cc(61)] Failed to post message to main thread.
```

I can confirm that changing the UIThreadPolicy via:

```
project.set_ui_thread_policy(flutter::UIThreadPolicy::RunOnSeparateThread);
```
sees these errors disappear. 

At this time I am not sure if it's an app code error or flutter engine error, but since this is a difference in the new default of a minor flutter version, I thought it best to reach out.

Do you have any guidance to narrow done the issue? Even with 'flutter run -v' there seems to be no other context to the error.

**Comment by knopp** [comment]
@robert-at-pieces, looks like https://github.com/flutter/flutter/issues/173843, which should be fixed by https://github.com/flutter/flutter/pull/176023. Can you try with current main and let me know if the problem persists?

**Comment by robert-at-pieces** [comment]
Unfortunately the issue persists using Flutter 3.37.0-1.0.pre-568 • channel master

I am currently trying to create a minimal reproducible example, but I have some (unverified) research with the help of an LLM. Might be related to flutter rust bridge and the changes with MessageLoops when the platform thread is merged with the UI thread.

[flutter_bug_report.md](https://github.com/user-attachments/files/22978413/flutter_bug_report.md)

**Comment by knopp** [comment]
If it persist it could be that you're simply running some long synchronous task on main thread. That's something you need to avoid.

**Comment by robert-at-pieces** [comment]
> If it persist it could be that you're simply running some long synchronous task on main thread. That's something you need to avoid.

Could be. I'll dig deeper. Thanks for the response.

**Comment by loic-sharma** [comment]
> If it persist it could be that you're simply running some long synchronous task on main thread. That's something you need to avoid.

@robert-at-pieces, You should be able to detect this using [Visual Studio's profiler](https://learn.microsoft.com/en-us/visualstudio/profiling/profiling-feature-tour?view=vs-2022&pivots=programming-language-cpp#analyze-cpu-usage) for native code, and DevTools CPU profiler for Dart code.

If you're still stuck after that, I'd consider updating Flutter Windows's `PostMessage` ([here](https://github.com/flutter/flutter/blob/4ddd558fab49f412acedae58683df1f0860948a5/engine/src/flutter/shell/platform/windows/task_runner_window.cc#L105) and [here](https://github.com/flutter/flutter/blob/4ddd558fab49f412acedae58683df1f0860948a5/engine/src/flutter/shell/platform/windows/task_runner_window.cc#L70)) to log `GetLastError` on failure to dump more information (see [this](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-postmessagea#remarks)). This will require doing your own build of the Flutter engine, so I'd do this as a last resort.

**Comment by squirelboy360** [comment]
With threads merged, are platform channels now comparable in performance to FFI 
(aside from the async vs sync behavior and serialization overhead)? 

In other words, was thread-hopping the main performance bottleneck, and now 
platform channels are "fast enough" for most use cases?

Is this similar to how React Native's bridgeless architecture improved 
performance by eliminating the async bridge, even though there's still some 
serialization overhead?

**Comment by stuartmorgan-g** [comment]
> are platform channels now comparable in performance to FFI (aside from the async vs sync behavior and serialization overhead)?

A performance comparison that omits key differences is unlikely to be useful in drawing real-world conclusions.

> was thread-hopping the main performance bottleneck, and now platform channels are "fast enough" for most use cases?

Platform channels were already fast enough for most use cases, as evidenced by the large and diverse array of plugins built using them. For the cases where they are not, the issues are—as is generally the case with performance issues—highly specific to the exact case. For some use cases, threading was largely irrelevant and copying data several times was prohibitive, and in other use cases data copies were irrelevant and thread hop latency and/or contention was a significant issue.

**Comment by ALLNL** [comment]
We're experiencing the same issue.
`[ERROR:flutter/shell/platform/windows/task_runner_window.cc(61)] Failed to post message to main thread.`

After downgrading from version 3.35.7 to 3.32.8, the issue was resolved.

**Comment by Flutter-DoHyunKim** [comment]
Hello, I upgraded Flutter from 3.24 to 3.35 to test the new merged Platform/UI threads.

After upgrading, I started seeing jank when scrolling a ListView that contains multiple PlatformViews on Android (iOS not tested yet).

Here is what I found:

- `DisableMergedPlatformUIThread: true` → Smooth scrolling 👍
- `DisableMergedPlatformUIThread: false` (default in 3.35) → Noticeable jank 👎

It seems that the merged thread model is causing performance regressions when rendering PlatformViews inside scrolling lists.

I understand that expensive logic should be moved to background threads. 
However, PlatformView rendering must run on the main thread (Android UI requirement), which is not something app developers can optimize.

This issue affects our production apps and module. 
Most ad SDKs (including Google's `google_mobile_ads`) use native views via PlatformView. Other widely used plugins like `google_maps_flutter` also rely on PlatformView.

The `DisableMergedPlatformUIThread` flag has been removed in Flutter 3.38, so we can no longer use it as a workaround. 
**Are there any recommended approaches to optimize PlatformView performance under the merged-thread architecture? 
Or is the Flutter team already working on improvements for PlatformView + merged-thread scenarios?**

Thank you.

**Comment by mit-mit** [comment]
> If you need a minimal reproduction repo, I can open a separate issue and share it!

Yes, please. Can you just upload it to somewhere (e.g. a Github gist) and share a link to it here?

**Comment by knopp** [comment]
@jacoplane, that should be fixed by https://github.com/flutter/flutter/pull/179538, can you check if your issue persists with current flutter main?

**Comment by srmncnk** [comment]
> As per Apple, we will need to update to Xcode 26 in April. And Flutter support for XCode 26 was added in 3.38 version, which no longer allows the opt-out flags.
> 
> <img width="704" height="115" alt="Image" src="https://github.com/user-attachments/assets/2f88885f-6b7b-4d17-a5f0-8e6511cc6bc4" /> 

 _Originally posted by @kaciula in [#169339](https://github.com/flutter/flutter/issues/169339#issuecomment-3951110121)_

Is this true? If so, it would be a major dealbreaker.

There's a promise for thread pinning as one of the alternatives for this (https://github.com/dart-lang/sdk/issues/46943) but I'm not sure it's going to make it until April (and even with that we'd still need time to rework our whole app architecture to make it work). The alternative proposal of having the flags be opt-in for headless engines does not seem to have traction: https://github.com/flutter/flutter/issues/176053.

If we turn this issue around - is there no way to get iOS 26 support for flutter versions, older than 3.38.x? There are some reports that it might work, I have not tested it yet myself:
> @kaciula 3.35 is compatible with xcode 26, I've pinned flutter to 3.35.7 without any SDK warnings.
> Another thing is UIScene migration https://docs.flutter.dev/release/breaking-changes/uiscenedelegate#migration-guide-for-flutter-apps but it will be a problem starting from 27 version I guess 

 _Originally posted by @p-shapovalov in [#169339](https://github.com/flutter/flutter/issues/169339#issuecomment-3951297859)_

**Comment by Gustl22** [comment]
Thank you for further advancing the integration of native platforms.
Unfortunately, currently there is no migration guide or example that shows how to deal with posting Tasks to the "main" thread for Windows and Linux. This is e.g. necessary to deal with events from the native plaform, which are emitted from another thread.

- No migration guide:
  - https://docs.flutter.dev/release/breaking-changes/macos-windows-merged-threads
  - https://docs.flutter.dev/release/breaking-changes/linux-merged-threads
- No easy way to post task on main thread: https://github.com/flutter/flutter/issues/79213
- No mention of windows or linux, how to jump to the main thread:
  - https://docs.flutter.dev/platform-integration/platform-channels#jumping-to-the-main-thread-in-ios
  - https://docs.flutter.dev/platform-integration/platform-channels#jumping-to-the-ui-thread-in-android
- No mention of how to deal with EventChannels: https://docs.flutter.dev/platform-integration/platform-channels#step-5-add-a-windows-platform-specific-implementation
- No example of EventChannels for Windows in https://github.com/flutter/packages/ (there either is no windows implementation, or one uses method_channel instead of event_channel to circumvent this problem)
- Examples for EventChannel, but without need of switching the thread: https://github.com/flutter/flutter/blob/a225be287e315e4bc5dd8ff9399fff6ac157f1a3/examples/platform_channel/windows/runner/flutter_window.cpp#L81

Did I overlook something? I would be glad, if you can advance more in EventChannels while forcing to a unified platform and UI thread in Windows and Linux.

We see this in our log and we would prefer a common way to do it used by all plugin maintainers:

```
[ERROR:flutter/shell/common/shell.cc(1064)] The 'xyz.luan/audioplayers/events/sfx-[numbers redacted]' channel sent a message from native to Flutter on a non-platform thread. Platform channel messages must be sent on the platform thread. Failure to do so may result in data loss or crashes, and must be fixed in the plugin or application code creating that channel.
See https://docs.flutter.dev/platform-integration/platform-channels#channels-and-platform-threading for more information.
```

Thank you for your help!

**Comment by Flutter-DoHyunKim** [comment]
Hi team,

We're planning to upgrade our Flutter add-to-app module from 3.24 → 3.41 in order to pick up the WebView gesture fix (PR #179908).

However, As previously discussed, we observed noticeable jank when scrolling PlatformView-heavy screens with merged Platform/UI threads enabled.

Since the opt-out flag was removed in PR #174408, we experimented with reverting that PR locally on top of the 3.41 stable engine to temporarily restore the non-merged thread behavior.

So far we were able to build and run the local engine successfully, and basic testing did not reveal any obvious issues.
The changes in that PR also seem relatively self-contained, but before considering this approach for production we wanted to confirm with the team.

Our intention would be to use this only as a temporary workaround, and switch back to the stable engine once PlatformView performance under merged threads improves 
(for example when HCPP lands or other related optimizations are available).

A couple of questions:

1. Are there any known risks with reverting PR #174408 and running the engine in non-merged thread mode on 3.41?
Would this be considered a reasonable temporary workaround for production?
2. Is there an active tracking issue or roadmap for improving PlatformView performance under merged threads?
Knowing roughly when this might improve would help us plan when we can remove the custom engine.

Thanks again for all the work on this.
Happy to provide more details or test builds if that would be helpful.

**Comment by Flutter-DoHyunKim** [comment]
@stuartmorgan-g 

Thanks for the detailed explanation. We understand the risks you mentioned around FFI packages that may rely on the merged-thread behavior in newer Flutter versions.

To reduce that risk, our current plan is to upgrade from 3.24 while keeping dependency updates minimal and avoiding packages that assume merged-thread behavior. In practice, we expect most of our dependencies to stay around the ~3.38 target range for now.

For the time being, we are considering a temporary custom engine workaround by reverting PR #174408 so that we can continue running with the non-merged thread configuration while evaluating newer Flutter versions.


While investigating the jank we observed, I did a bit of root cause analysis around PlatformView behavior under merged threads. During scrolling, setOffset appears to be called every frame for each visible PlatformView, which seems to trigger setLayoutParams() → requestLayout() → draw() repeatedly. With merged threads, this work appears to run serially on the main thread, which may explain the visible jank when several PlatformViews are on screen.

I also experimented with a small engine-side modification that avoids calling setLayoutParams() on every scroll-driven offset update and instead updates the wrapper position directly via layout(). This avoids the expensive requestLayout() path during scrolling, while still preserving the normal redraw flow when the embedded view content actually changes. In our test case, this removed the scrolling jank.

If this seems worth exploring, I'm happy to open a separate issue with more detailed analysis and the patch.

cc. @knopp @mit-mit @jtmcdole

**Comment by loic-sharma** [comment]
@Flutter-DoHyunKim Yes please file a separate issue for the Android platform view jank issue. I'd make sure to include a minimal Flutter app that reproduces the jank.

**Comment by stuartmorgan-g** [comment]
Is there an issue tracking removing the option on macOS? I have some plugin work that's blocked on being able to rely on merged threads for both iOS and macOS, so it would be helpful to be able to track the later.

**Comment by knopp** [comment]
> Is there an issue tracking removing the option on macOS? I have some plugin work that's blocked on being able to rely on merged threads for both iOS and macOS, so it would be helpful to be able to track the later.

Not that I'm aware of. We should remove it, it's not really tested or supported at this point on desktop platforms.

**Comment by stuartmorgan-g** [comment]
Turns out there is already https://github.com/flutter/flutter/issues/181874 which I missed before.

**Comment by svprdga** [comment]
The thread merge caused a serious performance problem with platform views on iOS. I detected
it in my apps that use AdMob banners, but it's reproducible with any platform view, even one
that sits outside the scrollable and never moves still makes an unrelated ListView janky.

I confirmed it's the merge: setting `FLTEnableMergedPlatformUIThread` to `false` on 3.35.7
makes it disappear completely, and setting it back to `true` brings it back.

[I've written a fully detailed report of the problem and how to reproduce it here.](https://github.com/flutter/flutter/issues/190815)

The bad part is that now that you can't opt out of the thread merge anymore, there's no way
to mitigate this.
